ir/src/constraints/graph.rs

```diff
@@ -165,7 +170,27 @@ impl AlgebraicGraph {
                     lhs.domain(),
                 ))
             }
-            Expression::ListFolding(_) => todo!(),
+            Expression::ListFolding(lf_type) => match lf_type {
+                ListFoldingType::Sum(lc) | ListFoldingType::Prod(lc) => {
+                    let list = unfold_lc(lc, symbol_table)?;
+                    let mut acc =
+                        self.insert_expr(symbol_table, &list[0], variable_roots, default_domain)?;
+                    for elem in list.iter().skip(1) {
+                        let expr =
+                            self.insert_expr(symbol_table, elem, variable_roots, default_domain)?;
+                        let op = match lf_type {
+                            ListFoldingType::Sum(_) => {
+                                Operation::Add(acc.root_idx(), expr.root_idx())
+                            }
+                            ListFoldingType::Prod(_) => {
+                                Operation::Mul(acc.root_idx(), expr.root_idx())
+                            }
+                        };
+                        acc = self.insert_bin_op(&acc, &expr, op)?;
+                    }
+                    Ok(acc)
+                }
```

Comment:
Let's move this into a separate function - something like `insert_list_folding()` (similar to how we have separate functions `insert_vector_access()` etc.).